client/network/src/discovery.rs

```diff
@@ -569,8 +583,19 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 							let ev = DiscoveryOut::UnroutablePeer(peer);
 							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(ev));
 						}
-						KademliaEvent::RoutablePeer { .. } | KademliaEvent::PendingRoutablePeer { .. } => {
-							// We are not interested in these events at the moment.
+						// TODO: By ignoring the `addresses` field we ignore the
+						// addresses reported through Kademlia. Are the
+						// addresses reported via the identify behaviour always
+						// superior? We depend on the identify behaviour
+						// finishing before the connection is closed as
+						// otherwise identify does not know how to connect to
+						// the node as we drop the addresses here.
```

Comment:
@tomaka is this something we can ignore? I would assume that (1) through the identify protocol itself a node knows best what its external addresses are and (2) the identify procedure succeeds before the connection initiated by Kademlia is closed.

Reply:
To me, the Kademlia addresses are superior to the addresses reported by `identify`.

As far as I understand, you are guaranteed that that the Kademlia addresses reported by `RoutablePeer` are always correct, and that the local node can actually connect to the target through this address.

Meanwhile the external addresses of `identify` are a "best effort" mechanism. The sender is free to report any address it wants or not report some addresses.

If two nodes are both behind a gateway, or on the same machine, or in some situation where IP addresses are non-global, then we know that these Kademlia addresses work while the `identify` ones might have some quirks.

Reply:
> As far as I understand, you are guaranteed that that the Kademlia addresses reported by RoutablePeer are always correct, and that the local node can actually connect to the target through this address.

Correct.

As mentioned above the local node and the remote node might be within a special subnet (e.g. behind a NAT) and thus, while the reported address through `RoutablePeer` might be routable for the local node, it might not be routable for nodes outside the before mentioned subnet.

With this in mind I would advocate for adding both the Kademlia reported address as well as the addresses reported through identify to the Kademlia DHT.

---

Calling `Kademlia::add_address` with the address reported via `KademliaEvent::RoutablePeer` after an `Identified` event forces the `Discovery` behaviour to keep track of the address for the time in between. It is just a small cache, but it is state that one needs to manage.

While the address reported through `KademliaEvent::RoutablePeer` is superior to the addresses reported through the identify behaviour, should it not be the case that the addresses reported through identify eventually contains that same address given that identify periodically exchanges observed addresses?

Reply:
> While the address reported through KademliaEvent::RoutablePeer is superior to the addresses reported through the identify behaviour, should it not be the case that the addresses reported through identify eventually contains that same address given that identify periodically exchanges observed addresses?

Eventually yes, but not necessarily. It also relates to #4920 

It feels a bit like a stupid situation. The small buffer you're talking about is already managed by libp2p, and all we would need is a way to obtain the information, but it's a bit complicated by the fact that we're indirectly inside a polling function.

Reply:
@romanb Do you have an opinion?

Reply:
In a way, I do feel that it would not be a bad idea for nodes to be able to be able to decide which of their addresses are inserted in the DHT.

Reply:
Do I understand correctly that you don't object to the current changes of this pull request @tomaka?

@romanb any objections?